### PR TITLE
Benchmark-user-guide

### LINKED ISSUE #34: Add user guide: Defining and parametrizing benchmarks by applying decorators
Add user guide to the documentation:
- [ ] Defining and parametrizing benchmarks by applying decorators (`@benchmark,@parametrize,@product)` - motivation, behavior, and one usage example for each.

### PR DESCRIPTION
**Pull request by maxmynter**
- Simplify the example in quickstart.md and align it with the usage example of Benchmarks.
- Add benchmarks example
- Remove overloading of the `@parametrize` and `@product` decorators
- Set `follow_wrapped=False` in `Interface.from_callable`.

Closes #34

### CODE REVIEW

**Comment by maxmynter** [comment]
Should we rename the `parameters` of `parameterize` to params to stay consistent with the `runner.run` (or vice versa)?

File: src/nnbench/core.py [hunk 1]
```diff
@@ -101,39 +101,7 @@ def decorator(fun: Callable) -> Benchmark:
...
-    parameters: Iterable[dict[str, Any]] = (),
-    setUp: Callable[..., None] = NoOp,
-    tearDown: Callable[..., None] = NoOp,
-    tags: tuple[str, ...] = (),
-) -> list[Benchmark]:
-    ...
-
-
-def parametrize(
-    func: Callable[..., Any] | None = None,
```
Comment (nicholasjng) [inline src/nnbench/core.py:136]: Please also remove the default tuple for `parameters` below.

File: docs/guides/benchmarks.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
+# Benchmarks
+
+You benchmark your machine learning code in nnbench by encapsulating the benchmarking code in separate functions and applying one of the decorators provided. 
+The available decorators are 
+- `@benchmark`, which runs a benchmark with supplied parameters,
+- `@parametrize`, which runs several benchmarks with the supplied parameter configurations,
+- `@product`, which runs benchmarks with all parameter combinations that arise from the supplied values.
```
Comment (nicholasjng) [inline docs/guides/benchmarks.md:7]: We might want to give all of them as absolute imports like `@nnbench.benchmark`, since that's what we have done so far in tests and examples.

File: docs/guides/benchmarks.md [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+You benchmark your machine learning code in nnbench by encapsulating the benchmarking code in separate functions and applying one of the decorators provided. 
+The available decorators are 
+- `@benchmark`, which runs a benchmark with supplied parameters,
+- `@parametrize`, which runs several benchmarks with the supplied parameter configurations,
+- `@product`, which runs benchmarks with all parameter combinations that arise from the supplied values. 
+
+We will first introduce a small machine learning example which we will subsequently use to showcase the motivation for and usage of the benchmarking decorators.
+
+We recommend to compartmentalize the logic of the machine learning model, the benchmarks, and benchmark execution into different files. In this guide, these are called `training.py`, `benchmarks.py`, and `run_benchmarks.py`.
```
Comment (nicholasjng) [inline docs/guides/benchmarks.md:11]:
```suggestion
We recommend to split the model training, benchmark definition, and benchmark running into different files. In this guide, these are called `training.py`, `benchmarks.py`, and `main.py`.
```

Then adjust all occurrences of `run_benchmarks.py` into main.py.

File: docs/guides/benchmarks.md [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+```python
+# training.py
+import numpy as np
+from sklearn import base, metrics
+from sklearn.datasets import load_iris
+from sklearn.ensemble import RandomForestClassifier
+from sklearn.model_selection import train_test_split
+
+
+def prepare_data() -> tuple[np.ndarray, np.ndarray,np.ndarray, np.ndarray]:
```
Comment (nicholasjng) [inline docs/guides/benchmarks.md:26]:
```suggestion
def prepare_data() -> tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
```

I guess this is not linted by ruff?

File: docs/guides/benchmarks.md [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    return model
+
+
+def accuracy(model: base.BaseEstimator, y_test: np.ndarray, y_pred: np.ndarray) -> float:
+    accuracy = metrics.accuracy_score(y_test, y_pred)
+    return accuracy
+```
+
+## `@benchmark` for single benchmarks
+Now, inside a new file called `benchmarks.py` we define our benchmarks.
```
Comment (nicholasjng) [inline docs/guides/benchmarks.md:45]:
```suggestion
Now, we define our benchmarks in a new file called `benchmarks.py`.
```

File: docs/guides/benchmarks.md [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+We first encapsulate the benchmark logic into a function, `benchmark_accuracy()` which prepares the data, trains a classifier, and lastly, obtains the accuracy.
```
Comment (nicholasjng) [inline docs/guides/benchmarks.md:46]: I mean, if we're doing this, we might as well make it the whole pipeline. Realistically, model training and preparation should not be a part of the function.

File: docs/guides/benchmarks.md [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+```bash
+python run_benchmarks.py  
+
+
+name         value
+--------  --------
+accuracy  0.933333
+```
+
+# `@parametrize` for multiple configuration benchmarks
```
Comment (nicholasjng) [inline docs/guides/benchmarks.md:86]:
```suggestion
## `@nnbench.parametrize` for multiple configuration benchmarks
```

(Please use double ## for subheadings.)

File: docs/guides/benchmarks.md [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+Sometimes, we are not only interested in the performance of a model for given parameters but want to compare the performance for different configurations. 
+To achieve this, let's modify our `benchmarks.py` to run a parametrized benchmark.
```
Comment (nicholasjng) [inline docs/guides/benchmarks.md:89]:
```suggestion
To achieve this, we can turn our single accuracy benchmark in the `benchmarks.py` file into a parametrized benchmark.
```

File: docs/guides/benchmarks.md [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+name         value
+--------  --------
+accuracy  0.933333
+```
+
+# `@parametrize` for multiple configuration benchmarks
+
+Sometimes, we are not only interested in the performance of a model for given parameters but want to compare the performance for different configurations. 
+To achieve this, let's modify our `benchmarks.py` to run a parametrized benchmark. 
+We need to replace the decorator with our `@parametrize` configuration and supply the wanted parameter combinations within an iterable.
```
Comment (nicholasjng) [inline docs/guides/benchmarks.md:90]:
```suggestion
To do this, replace the decorator with `@nnbench.parametrize` and supply the parameter combinations of choice as dictionaries in the first argument.
```

File: docs/guides/benchmarks.md [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+import nnbench
+from training import prepare_data, train_rf, accuracy
+
+@nnbench.parametrize((
+    {"n_estimators": 10, "max_depth": 2},
+    {"n_estimators": 50, "max_depth": 5},
+    {"n_estimators": 100, "max_depth": 10}
+))
+def benchmark_accuracy(n_estimators: int, max_depth: int, random_state: int) -> float:
+    ... #   Our benchmarking logic from above
```
Comment (nicholasjng) [inline docs/guides/benchmarks.md:103]: Please paste the logic again here.

File: docs/guides/benchmarks.md [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+```
+
+We still need to supply the remaining arguments to the `BenchmarkRunner()`, in our case the `random_state`.
```
Comment (nicholasjng) [inline docs/guides/benchmarks.md:106]: I would like a small sentence before this a la "Notice that the parametrization is incomplete, since we did not specify a value for the `random_state` argument".

Then, just name the argument, i.e. "We now need to supply a value for the `random_state` argument of our benchmark."

File: docs/guides/benchmarks.md [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+
+
+name                                                 value
+------------------------------------------------  --------
+benchmark_accuracy_n_estimators=10_max_depth=2    0.955556
+benchmark_accuracy_n_estimators=50_max_depth=5    0.866667
+benchmark_accuracy_n_estimators=100_max_depth=10  0.911111
+```
+
+# `@product` for benchmark scans over parameter configurations
```
Comment (nicholasjng) [inline docs/guides/benchmarks.md:129]:
```suggestion
## `@nnbench.product` for benchmarks over parameter configuration grids
```

File: docs/guides/benchmarks.md [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+
+In case we want to run a benchmark scan for all possible parameter combinations we can use the `@product` decorator and supply the different values for each parameter.
```
Comment (nicholasjng) [inline docs/guides/benchmarks.md:131]:
```suggestion
In case we want to run a benchmark scan for all possible combinations of a set of parameters, we can use the `@nnbench.product` decorator to supply the different values for each parameter. 
```

File: docs/guides/benchmarks.md [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+# benchmarks.py
+import nnbench
+from training import prepare_data, train_rf, accuracy
+
+@nnbench.product(n_estimators=[10, 50, 100], max_depth=[2, 5, 10])
+def benchmark_accuracy_product(n_estimators: int, max_depth: int, random_state: int) -> float:
+    ... #   Our benchmarking logic from above
+```
+
+We still provide the `random_state` parameter to the runner like we did with the `@parametrize` decorator.
```
Comment (nicholasjng) [inline docs/guides/benchmarks.md:143]:
```suggestion
We still provide the `random_state` parameter to the runner directly, like we did with the `@nnbench.parametrize` decorator.
```

File: docs/guides/benchmarks.md [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+In case we want to run a benchmark scan for all possible parameter combinations we can use the `@product` decorator and supply the different values for each parameter. 
+
+```python
+# benchmarks.py
+import nnbench
+from training import prepare_data, train_rf, accuracy
+
+@nnbench.product(n_estimators=[10, 50, 100], max_depth=[2, 5, 10])
+def benchmark_accuracy_product(n_estimators: int, max_depth: int, random_state: int) -> float:
+    ... #   Our benchmarking logic from above
```
Comment (nicholasjng) [inline docs/guides/benchmarks.md:140]: See above, add the code.

**Review by nicholasjng — COMMENTED** [review]
Sweet, this is starting to look like the real deal!

File: docs/guides/benchmarks.md [hunk 16]
```diff
@@ -0,0 +1,175 @@
+# Benchmarks
```
[the remark this answers is not in the document]
Comment (nicholasjng) [reply docs/guides/benchmarks.md:1]: It should also be the first page in the user guide since it's so important (first entry in the mkdocs.yml user guide list), so I will wait for this to be merged and rebase before checking in the organization best practices guide.
